### DISCUSSION TITLE
Large graph 2-hop sampling

### DISCUSSION THREAD

**Discussion by brovatten** [pyg-team/pytorch_geometric#7786] [Q&A] [answered]
Hi,

I want to offline sample 2-hop neighbours for every node of my Wikidata subset (5M nodes, 100M edges). This has to be _without batching_. Ultimately I want a dictionary with each node and its 2-hop neighbourhood.

It’s important that all 1-hop neighbours share all its possible edges with eachother (including 2-hop edges hence requires 2-hop node sampling I think). Remaining 2-hop nodes are not as important, but would optimally also be contained.

The problem is NeighborLoader only samples 1hop, and NeighborSampler is too slow. I would need to run it on GPU or use something custom. I tried doing it manually with tensors but it also took too much time.

Thesis is closing in so appreciate all advice 😄

**Comment by brovatten** [comment]
For reference, this is my manual code. I need to speed it up by about >3 times. 


```
adj_matrix = torch.sparse_coo_tensor(indices=a, values =torch.ones(a.shape[1], dtype=torch.long), device="cuda:0").coalesce()
unique_nodes = torch.unique(adj_matrix.indices()).tolist()

#unique_nodes = unique_nodes[:10000]
idx2subgraph = {}

for node in unique_nodes:
    onehops_nodes = adj_matrix[node].coalesce().indices()[0]

    mask = torch.isin(a[0], onehops_nodes) & torch.isin(a[1], onehops_nodes)
    edges = a[:, mask]
    idx2subgraph[node] = edges

    if len(idx2subgraph) % 10000 == 0:
        with open(pickle_path, "ab") as f:
            pickle.dump(idx2subgraph, f)
        idx2subgraph.clear()

```

**Comment by wsad1** [comment, marked as answer]
Why do you say `NeighborLoader` only samples 1-hop neighbors . 
[NeighborLoader](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.NeighborLoader) can sample 2-hop neighbours too via the `num_neighbors` argument. To sample all 1-hop and 2-hop nodes set `num_neighbors=[-1, -1]`,

**Reply by brovatten** [reply]
Great this solves it thank you! I was confused since neighboursampler used `sizes` instead, I assumed `num_neighbors` was an integer value not a list. 

It was much faster than my manual computation!

**Reply by brovatten** [reply]
If I want to do this online, in a multimodal setting such that I don't know which order I will need the links in, can I sample it efficiently on the fly or should I store it preemptively? E.g. subgraph(node_id) for a undefined stream of node_ids. 

Efficiency is my main concern as number of links are roughly 50 million.

**Reply by rusty1s** [reply]
You should be able to do this online, e.g., by overriding the `sampler` argument in the `DataLoader`.

**Reply by brovatten** [reply]
Hi Mattias. Thanks a lot for your reply. I tried doing this and it seems to be working. 

I have one last question: I am using a graph model that requires me to extract the unbatched subgraph for a node. However, setting batch size to 1 is so incredibly much slower than 32 and thus infeasible. So, is there any very efficient way I can batch the subgraphs, resulting in roughly size (2,6000) and then extract the 2-hop neighbours for each of the input nodes? Or is there a better way?

This is my settings: 
```NeighborLoader(data=data, batch_size=32, num_neighbors=[100,1],is_sorted=True, input_nodes= subsample, directed=False, shuffle=False, num_workers=3)```

**Reply by rusty1s** [reply]
You can use the `disjoint=True` option. This way, you should be able to extract each individual subgraph via the `batch` argument.

**Reply by brovatten** [reply]
Great idea, it seems to be working. Thank you so much!
